File: 2023-12-19/1759.js
const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : __dirname + '/input.txt';
const [L, data] = fs.readFileSync(filePath).toString().trim().split('\n');

const solution = ([n, m], sList) => {
  const pwdList = [];
  sList.sort();

  const checkPwd = (combi) => {
    let [consonants, vowels] = [0, 0];

    for (let i = 0; i < combi.length; i++) {
      if (['a', 'e', 'i', 'o', 'u'].includes(combi[i])) {
        vowels++;
      } else {
        consonants++;
      }
    }
    if (consonants >= 2 && vowels >= 1) {
      return true;
    } else {
      return false;
    }
  };

  const backtracking = (combi) => {
    if (combi.length === n) {
      if (checkPwd(combi)) {
        pwdList.push(combi.join(''));
        return;
      }
    }

    for (let idx = 0; idx < m; idx++) {
      if (combi[combi.length - 1] < sList[idx]) {
        combi.push(sList[idx]);
        backtracking(combi);
        combi.pop();
      }
    }
  };

  for (let i = 0; i <= m - n; i++) {
    backtracking([sList[i]]);
  }
  console.log(pwdList.join('\n'));
};

solution(L.split(' ').map(Number), data.split(' '));

File: 2023-12-19/1759.py
# 백트래킹을 이용한 풀이
L, C = list(map(int, input().split()))
sList = sorted(input().split())
pwdList = []

def checkPwd(combi):
    consonants, vowels = 0, 0
    for s in combi:
        if s in ['a', 'e', 'i', 'o', 'u']:
            vowels += 1
        else:
            consonants += 1
    if consonants >= 2 and vowels >= 1:
        return True
    else:
        return False

def backtracking(combi):
    if len(combi) == L:
        if checkPwd(combi):
            pwdList.append(''.join(combi))
            return
    
    for idx in range(len(combi), C):
        if combi[-1] < sList[idx]:
            combi.append(sList[idx])
            backtracking(combi)
            combi.pop()

for i in range(C - L + 1):
    backtracking([sList[i]])
    
print('\n'.join(pwdList))

# 조합을 이용한 풀이
from itertools import combinations

L, C = list(map(int, input().split()))
combiList = list(combinations(sorted(input().split()), L))
pwdList = []

for combi in combiList:
    consonants, vowels = 0, 0
    for s in combi:
        if s in ['a', 'e', 'i', 'o', 'u']:
            vowels += 1
        else:
            consonants += 1
    if consonants >= 2 and vowels >= 1:
        pwdList.append(''.join(combi))

print('\n'.join(pwdList))